src/zarr/v3/sync.py

```diff
@@ -113,7 +113,7 @@ def _sync(self, coroutine: Coroutine[Any, Any, T]) -> T:
     def _sync_iter(self, coroutine: Coroutine[Any, Any, AsyncIterator[T]]) -> List[T]:
         async def iter_to_list() -> List[T]:
             # TODO: replace with generators so we don't materialize the entire iterator at once
-            async_iterator = await coroutine
-            return [item async for item in async_iterator]
+            # async_iterator = await coroutine
+            return [item async for item in coroutine()]
```

Comment:
@DahnJ I had to revert some of your changes from #1692 here in order to get the code to run. The type signatures are currently incorrect, but the code works, so we should either adjust the annotations to match the actual types, or adjust the implementation (and the actual types) to match the annotation.

@jhamman, I know you have some ideas about the direction for this